text/0000-erfc-post-build-contexts.md

```diff
@@ -0,0 +1,416 @@
+
+A major drawback of this proposal is that it is very general, and perhaps too powerful. We're currently using the
+more focused API in the eRFC, and may switch to this during experimentation if a pressing need crops up.
+
+## Alternative procedural macro with minimal compiler changes
+
+The above proposal can be made even more general, minimizing the impact on the compiler.
+
+This assumes that `#![foo]` ("inner attribute") macros work on modules and on crates.
+
+The idea is that the compiler defines no new proc macro surface, and instead simply exposes
+a `--attribute` flag. This flag, like `-Zextra-plugins`, lets you attach a proc macro attribute
+to the whole crate before compiling. (This flag actually generalizes a bunch of flags that the
+compiler already has)
+
+Test crates are now simply proc macro attributes:
+
+```rust
+#[test_framework(attributes(test, foo, bar))]
+pub fn harness(crate: TokenStream) -> TokenStream {
+  // ...
+}
+```
+
+The cargo functionality will basically compile the file with the right dependencies
+and `--attribute=your_crate::harness`.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+These are mostly intended to be resolved during the experimental
+feature. Many of these have strawman proposals -- unlike the rest of this RFC,
+these proposals have not been discussed as thoroughly. If folks feel like
+there's consensus on some of these we can move them into the main RFC.
+
+## Integration with doctests
+
+Documentation tests are somewhat special, in that they cannot easily be
+expressed as `TokenStream` manipulations. In the first instance, the
+right thing to do is probably to have an implicitly defined framework
+ called `doctest` which is included in the testing set
+`test` by default (as proposed above).
+
+Another argument for punting on doctests is that they are intended to
+demonstrate code that the user of a library would write. They're there
+to document *how* something should be used, and it then makes somewhat
+less sense to have different "ways" of running them.
+
+## Standardizing the output
+
+We should probably provide a crate with useful output formatters and
+stuff so that if test harnesses desire, they can use the same output
+formatting as a regular test. This also provides a centralized location
+to standardize things like json output and whatnot.
+
+## Namespacing
+
+Currently, two frameworks can both declare interest in the same
+attributes. How do we deal with collisions (e.g., most test crates will
+want the attribute `#[test]`). Do we namespace the attributes by the
+framework name (e.g., `#[mytest::test]`)? Do we require them to be behind
+`#[cfg(mytest)]`?
+
+## Runtime dependencies and flags
+
+The code generated by the framework may itself have dependencies.
+Currently there's no way for the framework to specify this. One
+proposal is for the crate to specify  _runtime_ dependencies of the
+framework via:
+
+```toml
+[testing.framework.dependencies]
```

Comment:
I'm weary of scoping these dependencies under `testing.framework`, as that's the key used by crates that *use* a CTF. Also, isn't `testing.framework` already an array (which is why `[[testing.framework]]` works in users)?